File: webapp/src/nfc/type2.ts
/**
 * NFC-Forum Type-2 tag helpers: NDEF-message TLV framing (written from page 4),
 * NTAG type detection, and per-type capacity math. Mirrors the Android app's
 * maxPayloadForNdefCapacity so the same data fits the same tags.
 */

import { NdefFormatError, encodeNdefMime } from './ndef.js';
import { TOTAL_OVERHEAD } from '../chunk.js';

export function wrapType2Tlv(ndef: Uint8Array): Uint8Array {
  if (ndef.length < 0xff) {
    const out = new Uint8Array(2 + ndef.length + 1);
    out[0] = 0x03;
    out[1] = ndef.length;
    out.set(ndef, 2);
    out[out.length - 1] = 0xfe;
    return out;
  }
  const out = new Uint8Array(4 + ndef.length + 1);
  out[0] = 0x03;
  out[1] = 0xff;
  out[2] = (ndef.length >>> 8) & 0xff;
  out[3] = ndef.length & 0xff;
  out.set(ndef, 4);
  out[out.length - 1] = 0xfe;
  return out;
}

export function readType2Ndef(memory: Uint8Array): Uint8Array {
  let i = 0;
  while (i < memory.length) {
    const tag = memory[i]!;
    if (tag === 0x00) { i += 1; continue; } // NULL TLV
    if (tag === 0xfe) break; // terminator
    if (i + 1 >= memory.length) break;
    let len = memory[i + 1]!;
    let valueStart = i + 2;
    if (len === 0xff) {
      len = (memory[i + 2]! << 8) | memory[i + 3]!;
      valueStart = i + 4;
    }
    if (tag === 0x03) {
      if (valueStart + len > memory.length) throw new NdefFormatError('NDEF TLV runs past end of tag memory');
      return memory.slice(valueStart, valueStart + len);
    }
    i = valueStart + len; // skip lock (0x01) / memory-control (0x02) / other TLVs
  }
  throw new NdefFormatError('No NDEF TLV found in tag memory');
}

export enum NtagType {
  NTAG213 = 'NTAG213',
  NTAG215 = 'NTAG215',
  NTAG216 = 'NTAG216',
}

const USER_BYTES: Record<NtagType, number> = {
  [NtagType.NTAG213]: 144,
  [NtagType.NTAG215]: 504,
  [NtagType.NTAG216]: 888,
};

export function ntagUserBytes(t: NtagType): number {
  return USER_BYTES[t];
}

/** Total pages per type INCLUDING the config/lock pages, not just user memory.
 *  A raw dump needs all of them; `USER_BYTES` above covers only the NDEF area. */
const TOTAL_PAGES: Record<NtagType, number> = {
  [NtagType.NTAG213]: 45,
  [NtagType.NTAG215]: 135,
  [NtagType.NTAG216]: 231,
};

export function ntagTotalPages(t: NtagType): number {
  return TOTAL_PAGES[t];
}

export function detectNtagType(getVersion: Uint8Array): NtagType | null {
  const storage = getVersion[6];
  if (storage === 0x0f) return NtagType.NTAG213;
  if (storage === 0x11) return NtagType.NTAG215;
  if (storage === 0x13) return NtagType.NTAG216;
  return null;
}

/** Largest NFAR chunk payload whose NDEF+TLV-wrapped form fits `capacity` bytes.
 *  `capacity` is the NDEF data area available for the whole TLV (T + L + record +
 *  terminator) — for a real tag that is the Capability-Container-declared area
 *  (see ndefCapacityFromCC), NOT the raw user memory, which is what Android
 *  enforces when it reads the tag. */
export function chunkPayloadForCapacity(capacity: number): number {
  const wrappedLen = (payload: number): number =>
    wrapType2Tlv(encodeNdefMime(new Uint8Array(TOTAL_OVERHEAD + payload))).length;
  // Search downward; overhead is < 80 bytes so this is cheap.
  for (let p = capacity; p > 0; p--) {
    if (wrappedLen(p) <= capacity) return p;
  }
  return 0;
}

/** Pre-tap estimate only: max chunk payload for a tag TYPE using its raw user
 *  memory. The authoritative per-card capacity comes from the tag's Capability
 *  Container at write time (ndefCapacityFromCC → chunkPayloadForCapacity), which
 *  can be smaller (NTAG215: CC 496 B vs 504 B raw; NTAG216: 872 vs 888). */
export function ntagChunkPayloadSize(t: NtagType): number {
  return chunkPayloadForCapacity(USER_BYTES[t]);
}

/**
 * The NFC-Forum Type-2 Capability Container is 4 bytes at page 3:
 * [0]=0xE1 magic, [1]=version, [2]=MLEN (memory size / 8), [3]=access.
 * The NDEF data area available for the TLV is MLEN × 8 bytes. Android reads a
 * tag bounded by this area, so it is the capacity the writer must respect.
 * Returns null when the CC is absent/invalid (not an NDEF-formatted tag).
 */
export function ndefCapacityFromCC(cc: Uint8Array): number | null {
  if (cc.length < 4 || cc[0] !== 0xe1 || cc[2] === 0) return null;
  return cc[2]! * 8;
}

/** NDEF data area each NTAG type declares in its FACTORY Capability Container
 *  (MLEN x 8). Smaller than raw user memory: NTAG215 496 vs 504, NTAG216 872 vs
 *  888, NTAG213 144 == 144. */
const FACTORY_CC_BYTES: Record<NtagType, number> = {
  [NtagType.NTAG213]: 144,
  [NtagType.NTAG215]: 496,
  [NtagType.NTAG216]: 872,
};

export function ntagFactoryNdefCapacity(t: NtagType): number {
  return FACTORY_CC_BYTES[t];
}

/** Chunk payload for a reader that cannot read the tag's own Capability
 *  Container — Web NFC. Assumes the factory CC, which is what an unmodified tag
 *  ships with, so the bytes match what the Chameleon path writes after reading
 *  the real CC. */
export function webNfcChunkPayload(t: NtagType): number {
  return chunkPayloadForCapacity(ntagFactoryNdefCapacity(t));
}

File: lib/core/nfc/type2.dart
import 'dart:typed_data';

import 'ndef_bytes.dart';

/// NFC-Forum Type-2 tag helpers: NDEF-message TLV framing (written from page
/// 4), NTAG type detection, and per-type geometry.
///
/// Port of `webapp/src/nfc/type2.ts`.

/// The page an NDEF TLV starts at on every NTAG. Pages 0–3 are UID, lock bytes
/// and the capability container.
const int ndefStartPage = 4;

/// The NTAG chips this app reads and writes.
enum NtagType { ntag213, ntag215, ntag216 }

/// Frame an NDEF message as a Type-2 NDEF TLV, terminator included.
///
/// The length field has two forms and the boundary is easy to get wrong: one
/// byte below 255, otherwise `0xFF` followed by a 16-bit length. Choosing the
/// short form at 255 would corrupt every card larger than an NTAG213, and it
/// is invisible on small test payloads.
Uint8List wrapType2Tlv(Uint8List ndef) {
  if (ndef.length < 0xff) {
    final out = Uint8List(2 + ndef.length + 1);
    out[0] = 0x03;
    out[1] = ndef.length;
    out.setRange(2, 2 + ndef.length, ndef);
    out[out.length - 1] = 0xfe;
    return out;
  }
  final out = Uint8List(4 + ndef.length + 1);
  out[0] = 0x03;
  out[1] = 0xff;
  out[2] = (ndef.length >> 8) & 0xff;
  out[3] = ndef.length & 0xff;
  out.setRange(4, 4 + ndef.length, ndef);
  out[out.length - 1] = 0xfe;
  return out;
}

/// Find and return the NDEF message inside Type-2 tag memory.
///
/// Walks the TLV chain, skipping NULL (0x00), lock-control (0x01) and
/// memory-control (0x02) entries, and stops at the terminator (0xFE).
///
/// Throws [NdefFormatException] while the TLV is still incomplete — which is
/// exactly what the inspector needs mid-dump: "no complete NDEF TLV yet" is a
/// different fact from "this tag holds no NFAR chunk".
Uint8List readType2Ndef(Uint8List memory) {
  var i = 0;
  while (i < memory.length) {
    final tag = memory[i];
    if (tag == 0x00) {
      i += 1;
      continue;
    }
    if (tag == 0xfe) break;
    if (i + 1 >= memory.length) break;

    var len = memory[i + 1];
    var valueStart = i + 2;
    if (len == 0xff) {
      if (i + 3 >= memory.length) break;
      len = (memory[i + 2] << 8) | memory[i + 3];
      valueStart = i + 4;
    }

    if (tag == 0x03) {
      if (valueStart + len > memory.length) {
        throw const NdefFormatException(
          'NDEF TLV runs past end of tag memory',
        );
      }
      return Uint8List.sublistView(memory, valueStart, valueStart + len);
    }
    i = valueStart + len;
  }
  throw const NdefFormatException('No NDEF TLV found in tag memory');
}

/// Bytes of *user* memory — the NDEF data area only.
const Map<NtagType, int> _userBytes = {
  NtagType.ntag213: 144,
  NtagType.ntag215: 504,
  NtagType.ntag216: 888,
};

/// Total pages per type **including** the config and lock pages, not just user
/// memory. A raw dump needs all of them; [ntagUserBytes] covers only the NDEF
/// area, and using it here would silently truncate every dump.
const Map<NtagType, int> _totalPages = {
  NtagType.ntag213: 45,
  NtagType.ntag215: 135,
  NtagType.ntag216: 231,
};

int ntagUserBytes(NtagType t) => _userBytes[t]!;

int ntagTotalPages(NtagType t) => _totalPages[t]!;

/// Identify the chip from an NTAG `GET_VERSION` (0x60) response.
///
/// Byte 6 is the storage-size code. Returns null for anything unrecognised —
/// callers turn that into an "unsupported tag" message that names the byte,
/// because for the card inspector that value is the finding.
NtagType? detectNtagType(Uint8List getVersion) {
  if (getVersion.length < 7) return null;
  switch (getVersion[6]) {
    case 0x0f:
      return NtagType.ntag213;
    case 0x11:
      return NtagType.ntag215;
    case 0x13:
      return NtagType.ntag216;
    default:
      return null;
  }
}

/// Human label for a chip, used in inspector output.
String ntagLabel(NtagType t) => switch (t) {
      NtagType.ntag213 => 'NTAG213',
      NtagType.ntag215 => 'NTAG215',
      NtagType.ntag216 => 'NTAG216',
    };
